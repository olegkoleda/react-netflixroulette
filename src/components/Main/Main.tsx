import React from "react";
import { StyledMain, StyledResults } from "./styled.main";
import { Composition, Box } from "atomic-layout";
import Filter from "../Filter";
import Sort from "../Sort";
import CardsList from "../CardsList";

const mockedData = {
  filters: [
    { filter: "All", checked: true },
    { filter: "Documentary", checked: false },
    { filter: "Comedy", checked: false },
    { filter: "Horror", checked: false },
  ],
  sort: ["Release date", "Rating", "Budget"],
};

const controlsAreasMd = `filter filter filter sort`;

const controlsAreasMobile = `
  filter
  sort
`;

const Main = () => (
  <Box as={StyledMain} paddingHorizontal={"1rem"} paddingHorizontalMd={"3rem"}>
    <Composition areas={controlsAreasMobile} areasMd={controlsAreasMd}>
      {(Areas) => (
        <>
          <Areas.Filter>
            <Filter filters={mockedData.filters} />
          </Areas.Filter>
          <Areas.Sort>
            <Sort sortValues={mockedData.sort} />
          </Areas.Sort>
        </>
      )}
    </Composition>
    <StyledResults>
      <span>39</span> movies found
    </StyledResults>
    <CardsList />
  </Box>
);

export default Main;
