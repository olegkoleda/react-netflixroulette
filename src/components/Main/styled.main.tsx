import styled from "styled-components";

export const StyledMain = styled.main`
  height: 100%;
  min-height: calc(50vh - 6.5rem);
  background: ${(props) => props.theme.color.black};
`;

export const StyledResults = styled.div`
  margin: 1.875rem 0;
  font-size: 1.125rem;

  & span {
    font-weight: 600;
  }
`;
