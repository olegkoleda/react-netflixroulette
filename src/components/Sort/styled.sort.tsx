import styled from "styled-components";

export const StyledSelectWrapper = styled.div`
  padding: 0.7rem;
  text-align: right;
`;

export const StyledLabel = styled.label`
  margin-right: 1rem;
  color: ${(props) => props.theme.colour.grey};
`;

export const StyledOption = styled.option`
  background: ${(props) => props.theme.colour.black};
  &:checked {
    background: ${(props) => props.theme.colour.red};
    color: ${(props) => props.theme.colour.white};
  }
`;

export const StyledSelect = styled.select`
  background: none;
  border: none;
  color: ${(props) => props.theme.colour.white};
  font-size: 1rem;
`;
