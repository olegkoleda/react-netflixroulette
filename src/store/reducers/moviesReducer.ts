import {
  LOAD_MOVIES_ERROR,
  LOAD_MOVIES_STARTED,
  LOAD_MOVIES_SUCCESS,
  SET_ACTIVE_MOVIE,
  LoadMovieActionType,
  ISetActiveMovie,
} from "../types/loadMovies";
import {
  ADD_MOVIE_ERROR,
  ADD_MOVIE_STARTED,
  ADD_MOVIE_SUCCESS,
  ADD_MOVIE_FINISHED,
  AddMovieActionType,
} from "../types/addMovie";
import {
  DELETE_MOVIE_ERROR,
  DELETE_MOVIE_STARTED,
  DELETE_MOVIE_SUCCESS,
  DeleteMovieActionType,
} from "../types/deleteMovie";
import { IMovie } from "../../interfaces/IMovie";
import {
  UPDATE_MOVIE_ERROR,
  UPDATE_MOVIE_STARTED,
  UPDATE_MOVIE_SUCCESS,
  UpdateMovieActionType,
  UPDATE_MOVIE_FINISHED,
} from "../types/updateMovie";

export interface IMoviesState {
  list: IMovie[];
  loading: boolean;
  isMovieUpdated: boolean;
  error: Error | null;
  activeMovie: number | null;
  isMovieAdded: boolean;
}

const initialState: IMoviesState = {
  list: [],
  loading: false,
  error: null,
  activeMovie: null,
  isMovieUpdated: false,
  isMovieAdded: false,
};

export default function moviesReducer(
  state = initialState,
  action:
    | LoadMovieActionType
    | ISetActiveMovie
    | AddMovieActionType
    | DeleteMovieActionType
    | UpdateMovieActionType
) {
  switch (action.type) {
    case LOAD_MOVIES_STARTED:
    case ADD_MOVIE_STARTED:
    case DELETE_MOVIE_STARTED:
    case UPDATE_MOVIE_STARTED:
      return {
        ...state,
        loading: true,
      };
    case LOAD_MOVIES_SUCCESS:
      return {
        ...state,
        loading: false,
        list: action.payload,
      };
    case LOAD_MOVIES_ERROR:
    case ADD_MOVIE_ERROR:
    case DELETE_MOVIE_ERROR:
    case UPDATE_MOVIE_ERROR:
      return {
        ...state,
        error: action.error,
      };
    case SET_ACTIVE_MOVIE:
      return {
        ...state,
        activeMovie: action.payload,
      };
    case UPDATE_MOVIE_FINISHED:
      return {
        ...state,
        isMovieUpdated: false,
      };
    case ADD_MOVIE_FINISHED:
      return {
        ...state,
        isMovieAdded: false,
      };
    case ADD_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieAdded: true,
        list: [action.payload, ...state.list],
      };
    case DELETE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        list: state.list.filter(({ id }) => id !== action.payload),
      };
    case UPDATE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieUpdated: true,
        list: state.list.map((movie) =>
          movie.id === action.payload.id ? action.payload : movie
        ),
      };
    default:
      return state;
  }
}
