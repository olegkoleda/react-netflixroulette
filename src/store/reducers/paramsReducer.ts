import { FilterMovieActionType, SET_MOVIE_FILTER } from "../types/filterMovie";

export interface IParamsState {
  filter: string;
  sortBy: string;
}

const initialState: IParamsState = {
  filter: "",
  sortBy: "release_date",
};

export default function paramsReducer(
  state = initialState,
  action: FilterMovieActionType
) {
  switch (action.type) {
    case SET_MOVIE_FILTER:
      return {
        ...state,
        filter: action.payload,
      };

    default:
      return state;
  }
}
