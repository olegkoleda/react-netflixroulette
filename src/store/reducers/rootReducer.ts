import { combineReducers } from "redux";
import { IParamsState } from "./paramsReducer";
import { IMoviesState } from "./moviesReducer";
import movies from "./moviesReducer";
import params from "./paramsReducer";

export default combineReducers({
  movies,
  params,
});

export interface IAppState {
  movies: IMoviesState;
  params: IParamsState;
}
