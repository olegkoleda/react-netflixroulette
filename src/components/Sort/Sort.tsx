import React, { useCallback } from "react";
import {
  StyledLabel,
  StyledOption,
  StyledSelect,
  StyledSelectWrapper,
} from "./styled.sort";
import { Box } from "atomic-layout";
import { connect } from "react-redux";
import { IAppState } from "../../store/reducers/rootReducer";
import { sortMovies } from "../../store/actions/sortMovies";
import { loadMovies } from "../../store/actions/loadMovies";

interface ISortType {
  value: string;
  label: string;
}
interface ISortProps {
  sortValues: ISortType[];
  selectedSorValue: string;
  setSortBy: Function;
  loadMoviesProp: Function;
}

const Sort = ({
  sortValues,
  selectedSorValue,
  setSortBy,
  loadMoviesProp,
}: ISortProps) => {
  const changeValue = useCallback((e: React.ChangeEvent<HTMLSelectElement>) => {
    setSortBy(e.target.value);
    loadMoviesProp();
  }, []);

  return (
    <Box
      as={StyledSelectWrapper}
      flex
      justifyContentMd={"flex-end"}
      alignItems={"center"}
    >
      <StyledLabel htmlFor="sort-by">SORT BY</StyledLabel>
      <StyledSelect
        id="sort-by"
        onChange={changeValue}
        defaultValue={selectedSorValue}
      >
        {sortValues.map(({ value, label }) => (
          <StyledOption key={`sort-${value}`} value={value}>
            {label.toUpperCase()}
          </StyledOption>
        ))}
      </StyledSelect>
    </Box>
  );
};

const mapState = (state: IAppState) => ({
  selectedSorValue: state.params.sortBy,
});

const mapDispatch = {
  setSortBy: (sort: string) => sortMovies(sort),
  loadMoviesProp: () => loadMovies(),
};

export default connect(mapState, mapDispatch)(Sort);
