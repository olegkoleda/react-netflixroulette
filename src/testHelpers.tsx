import React from "react";
import renderer from "react-test-renderer";
import { theme } from "./styles/Theme";
import { ThemeProvider } from "styled-components";
import { BrowserRouter as Router } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { initialAppState } from "./store/reducers/rootReducer";

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
const store = mockStore(initialAppState);

export function renderWithTheme(component: React.ReactNode) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export const reduxWrapper = (Component: typeof React.Component | React.FC, data = {}) => (
  <Provider store={store}>
    <Router>
      <Component {...data}/>
    </Router>
  </Provider>
);
