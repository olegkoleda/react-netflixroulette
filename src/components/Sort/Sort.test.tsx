import React from "react";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import Sort from "./Sort";
import { renderWithTheme } from "../../testHelpers";
import { initialAppState } from "../../store/reducers/rootReducer";
import { mockedData } from "../Main/Main";

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
const store = mockStore(initialAppState);

test("Sort component matches the Snapshot", () => {
  const component = renderWithTheme(
    <Provider store={store}>
        <Sort sortValues={mockedData.sort} />
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
