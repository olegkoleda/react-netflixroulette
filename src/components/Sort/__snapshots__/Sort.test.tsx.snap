// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort component renders correctly 1`] = `
<div
  className="styledsort__StyledSelectWrapper-sc-13rug2s-0 cQsuLb sc-AxjAm kDfEX"
>
  <label
    className="styledsort__StyledLabel-sc-13rug2s-1 iVKDWt"
    htmlFor="sort-by"
  >
    SORT BY
  </label>
  <select
    className="styledsort__StyledSelect-sc-13rug2s-3 iacsnF"
    defaultValue="release_date"
    id="sort-by"
    onChange={[Function]}
  >
    <option
      className="styledsort__StyledOption-sc-13rug2s-2 kiNmiY"
      value="release_date"
    >
      RELEASE DATE
    </option>
    <option
      className="styledsort__StyledOption-sc-13rug2s-2 kiNmiY"
      value="rating"
    >
      RATING
    </option>
    <option
      className="styledsort__StyledOption-sc-13rug2s-2 kiNmiY"
      value="budget"
    >
      BUDGET
    </option>
  </select>
</div>
`;
